### ISSUE TITLE
[onert] Segfault in sgd optimizer

### ISSUE THREAD

**Issue by jyoungyun** [Samsung/ONE#11860]
### What

When I train a model with a softmax layer, a segfault occurs.
This is related to #11658 PR.

<details>
<summary>backtrace</summary>

```
#10 0x00007ffff6436437 in std::_Function_handler<void (long, long), Eigen::internal::TensorExecutor<Eigen::TensorAssignOp<Eigen::TensorMap<Eigen::Tensor<float, 1, 1, long>, 16, Eigen::MakePointer>, Eigen::TensorCwiseBinaryOp<Eigen::internal::scalar_difference_op<float, float>, Eigen::TensorMap<Eigen::Tensor<float, 1, 1, long>, 16, Eigen::MakePointer> const, Eigen::TensorCwiseUnaryOp<Eigen::internal::bind2nd_op<Eigen::internal::scalar_product_op<float const, float const> >, Eigen::TensorMap<Eigen::Tensor<float const, 1, 1, long>, 16, Eigen::MakePointer> const> const> const> const, Eigen::ThreadPoolDevice, true, (Eigen::internal::TiledEvaluation)0>::run(Eigen::TensorAssignOp<Eigen::TensorMap<Eigen::Tensor<float, 1, 1, long>, 16, Eigen::MakePointer>, Eigen::TensorCwiseBinaryOp<Eigen::internal::scalar_difference_op<float, float>, Eigen::TensorMap<Eigen::Tensor<float, 1, 1, long>, 16, Eigen::MakePointer> const, Eigen::TensorCwiseUnaryOp<Eigen::internal::bind2nd_op<Eigen::internal::scalar_product_op<float const, float const> >, Eigen::TensorMap<Eigen::Tensor<float const, 1, 1, long>, 16, Eigen::MakePointer> const> const> const> const&, Eigen::ThreadPoolDevice const&)::{lambda(long, long)#1}>::_M_invoke(std::_Any_data const&, long&&, std::_Any_data const&) (__functor=..., __args#0=@0x7fffffffc7d0: 0, 
    __args#1=@0x7fffffffc7c8: 15616) at /usr/include/c++/9/bits/std_function.h:300
#11 0x00007ffff6424e1f in std::function<void (long, long)>::operator()(long, long) const (this=0x7fffffffcae0, __args#0=0, __args#1=15616) at /usr/include/c++/9/bits/std_function.h:688
#12 0x00007ffff6422c18 in Eigen::ThreadPoolDevice::parallelFor(long, Eigen::TensorOpCost const&, std::function<long (long)>, std::function<void (long, long)>) const::{lambda(long, long)#1}::operator()(long, long) const (__closure=0x5555556b9c00, firstIdx=0, lastIdx=15616)
    at /home/jyoung/git/npu/ONE/externals/TENSORFLOW-2.8.0-EIGEN/unsupported/Eigen/CXX11/src/Tensor/TensorDeviceThreadPool.h:226
#13 0x00007ffff64290ac in std::_Function_handler<void (long, long), Eigen::ThreadPoolDevice::parallelFor(long, Eigen::TensorOpCost const&, std::function<long (long)>, std::function<void (long, long)>) const::{lambda(long, long)#1}>::_M_invoke(std::_Any_data const&, long&&, std::_Any_data const&) (__functor=..., __args#0=@0x7fffffffc900: 0, __args#1=@0x7fffffffc8f8: 100352)
    at /usr/include/c++/9/bits/std_function.h:300
--Type <RET> for more, q to quit, c to continue without paging--
#14 0x00007ffff6424e1f in std::function<void (long, long)>::operator()(long, long) const (this=0x7fffffffc9a0, __args#0=0, __args#1=100352) at /usr/include/c++/9/bits/std_function.h:688
#15 0x00007ffff6422eda in Eigen::ThreadPoolDevice::parallelFor(long, Eigen::TensorOpCost const&, std::function<long (long)>, std::function<void (long, long)>) const (this=0x55555569cb30, 
    n=100352, cost=..., block_align=..., f=...) at /home/jyoung/git/npu/ONE/externals/TENSORFLOW-2.8.0-EIGEN/unsupported/Eigen/CXX11/src/Tensor/TensorDeviceThreadPool.h:233
#16 0x00007ffff6434fc5 in Eigen::internal::TensorExecutor<Eigen::TensorAssignOp<Eigen::TensorMap<Eigen::Tensor<float, 1, 1, long>, 16, Eigen::MakePointer>, Eigen::TensorCwiseBinaryOp<Eigen::internal::scalar_difference_op<float, float>, Eigen::TensorMap<Eigen::Tensor<float, 1, 1, long>, 16, Eigen::MakePointer> const, Eigen::TensorCwiseUnaryOp<Eigen::internal::bind2nd_op<Eigen::internal::scalar_product_op<float const, float const> >, Eigen::TensorMap<Eigen::Tensor<float const, 1, 1, long>, 16, Eigen::MakePointer> const> const> const> const, Eigen::ThreadPoolDevice, true, (Eigen::internal::TiledEvaluation)0>::run (expr=..., device=...) at /home/jyoung/git/npu/ONE/externals/TENSORFLOW-2.8.0-EIGEN/unsupported/Eigen/CXX11/src/Tensor/TensorExecutor.h:340
#17 0x00007ffff6434ac8 in Eigen::TensorDevice<Eigen::TensorMap<Eigen::Tensor<float, 1, 1, long>, 16, Eigen::MakePointer>, Eigen::ThreadPoolDevice>::operator-=<Eigen::TensorCwiseUnaryOp<Eigen::internal::bind2nd_op<Eigen::internal::scalar_product_op<float const, float const> >, Eigen::TensorMap<Eigen::Tensor<float const, 1, 1, long>, 16, Eigen::MakePointer> const> > (
    this=0x7fffffffcc30, other=...) at /home/jyoung/git/npu/ONE/externals/TENSORFLOW-2.8.0-EIGEN/unsupported/Eigen/CXX11/src/Tensor/TensorDevice.h:61
#18 0x00007ffff6434706 in nnfw::cker::training_ops::functor::ApplyGradientDescent<Eigen::ThreadPoolDevice, float>::operator()(Eigen::ThreadPoolDevice const&, Eigen::TensorMap<Eigen::Tensor<float, 1, 1, long>, 16, Eigen::MakePointer>, Eigen::TensorMap<Eigen::TensorFixedSize<float const, Eigen::Sizes<>, 1, long>, 16, Eigen::MakePointer>, Eigen::TensorMap<Eigen::Tensor<float const, 1, 1, long>, 16, Eigen::MakePointer>) (this=0x7fffffffccb7, d=..., var=..., lr=..., grad=...) at /home/jyoung/git/npu/ONE/compute/cker/include/cker/eigen/training_ops.h:158
#19 0x00007ffff643447f in nnfw::cker::train::GradientDescent (output_shape=..., output_data=0x7ffff6576038, grad_shape=..., grad_data=0x7ffff55ec038, learning_rate=0.00999999978)
    at /home/jyoung/git/npu/ONE/compute/cker/include/cker/train/optimizer/SGD.h:54
#20 0x00007ffff6433a82 in onert::backend::train::optimizer::SGD::applyGradient (this=0x55555569bb00, factors=std::tuple containing = {...})
--Type <RET> for more, q to quit, c to continue without paging--
    at /home/jyoung/git/npu/ONE/runtime/onert/backend/train/optimizer/SGD.cc:53
#21 0x00007ffff64181b5 in onert::backend::train::ops::GradientApplier::applyGradient (this=0x5555556b5920, training_step=0)
    at /home/jyoung/git/npu/ONE/runtime/onert/backend/train/ops/GradientApplier.cc:45
#22 0x00007ffff6ffa703 in onert::exec::train::TrainableFnSequence::backward (this=0x5555556b5c80, training_step=0)
```

</details>

**Comment by jyoungyun** [comment]
I couldn't attach the model file. Please contact me if you need a model file.

**Comment by jyoungyun** [comment]
f 0
```
/* Load four SPFP values from P.  The address must be 16-byte aligned.  */
extern __inline __m128 __attribute__((__gnu_inline__, __always_inline__, __artificial__))
_mm_load_ps (float const *__P)
{
  return *(__m128 *)__P;
}
```
```
(gdb) p __P
$5 = (const float *) 0x7ffff55ec038
```

f 19
```
  const training_ops::CPUDevice &device = *eigen_support::GetThreadPoolDevice();
  training_ops::functor::ApplyGradientDescent<training_ops::CPUDevice, float>()(
    device, output_tensor.flat<float>(), lr_tensor.scalar<float>(),
    static_cast<const Tensor &>(grad_tensor).flat<float>());
```
```
(gdb) p grad_tensor
$6 = {shape = {static kMaxSmallSize = 5, _size = 2, {_dims = {128, 784, -12864, 32767, -163638943}, _dims_pointer = 0x31000000080}}, buffer = 0x7ffff55ec038}
```

- grad_tensor generates segfault.

**Comment by jyoungyun** [comment]
> You are not ensuring that your data is 16 byte aligned (malloc/new are not sufficient in general) - you will either need to use _mm_loadu_ps instead of _mm_load_ps to deal with your potentially misaligned data, or preferably use a suitable method to allocate aligned memory (e.g. [posix_memalign](http://linux.die.net/man/3/posix_memalign) on Linux).
>
> from https://stackoverflow.com/questions/21015961/mm-load-ps-caused-segment-fault

```
0x7ffff6576038
(gdb) p trainable_tensor
$8 = {shape = {static kMaxSmallSize = 5, _size = 2, {_dims = {128, 784, -13024, 32767, -151260908}, _dims_pointer = 0x31000000080}}, buffer = 0x7ffff6576038}
```
```
0x7ffff55ec038
(gdb) p grad_tensor
$2 = {shape = {static kMaxSmallSize = 5, _size = 2, {_dims = {128, 784, -12896, 32767, -163638943}, _dims_pointer = 0x31000000080}}, buffer = 0x7ffff55ec038}
```
```
  float *temp = static_cast<float*>(_mm_malloc(grad_tensor.total_size(), 16));
  memcpy(temp, grad_tensor.buffer(), grad_tensor.total_size());

  float *temp2 = static_cast<float*>(_mm_malloc(trainable_tensor.total_size(), 16));
  memcpy(temp2, trainable_tensor.buffer(), trainable_tensor.total_size());

  switch (grad_tensor.data_type())
  {
    case ir::DataType::FLOAT32:
      nnfw::cker::train::GradientDescent(
        ops::getShape(&trainable_tensor), temp2,
        ops::getShape(&grad_tensor), temp, lr);
      break;
    default:
      throw std::runtime_error("SGD: Not supported data type");
  }
```

This code resolves the segfault error.
